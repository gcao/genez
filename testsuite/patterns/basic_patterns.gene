#!/usr/bin/env gene run

# Test: Basic pattern matching with literals, variables, and wildcards

# Literal patterns
(fn describe_number [n]
  (match n
    (0 "zero")
    (1 "one") 
    (2 "two")
    (_ "other")))

(print "Literal patterns:")
(print (describe_number 0))
(print (describe_number 1))
(print (describe_number 2))
(print (describe_number 99))

# Variable binding patterns
(fn process_value [x]
  (match x
    (0 "got zero")
    (n "got a value")))

(print "")
(print "Variable patterns:")
(print (process_value 0))
(print (process_value 42))

# Combining patterns
(fn categorize [x]
  (match x
    (0 "zero")
    (1 "one")
    (n (if (< n 10)
         "small number"
       else
         "large number"))))

(print "")
(print "Combined patterns:")
(print (categorize 0))
(print (categorize 1))
(print (categorize 5))
(print (categorize 100))