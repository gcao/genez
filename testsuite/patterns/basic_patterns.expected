Literal patterns:
zero
one
two
other

Variable patterns:
got zero
got a value

Combined patterns:
zero
one
small number
large number